A98-SNI-setting-and-SNI-SAN-validation.md

```diff
@@ -0,0 +1,117 @@
+channel arguments or is put in sub-channel attribute wrapped in a
+`SslContextProviderSupplier`, depending on the language. The `UpstreamTlsContext.SNI`
+would already be available to this provider supplier  from the parsed Cluster resource.
+At the time of Tls protocol negotiation, when this provider supplier is 
+used to invoked to set the SslContext, the hostname from the channel attributes
+also will be passed now, to determine the SNI to be set for the Tls handshake.
+For example, in Java, at protocol negotiation time the `SslContextProviderSupplier` is given 
+a callback to be invoked with the `SslContext` when the client Ssl Provider instantiated by 
+this supplier has the `SslContext` ready. This callback will now also be passed the SNI
+taken from the subchannel attributes. This value along with the `UpstreamTlsContext` available
+in the `SslContextProviderSupplier` will be used to decide the SNI to be used for the handshake.
+
+[A29_impl-details]: https://github.com/grpc/proposal/blob/master/A29-xds-tls-security.md#implementation-details
+[UTC_SNI]: https://github.com/envoyproxy/envoy/blob/ee2bab9e40e7d7649cc88c5e1098c74e0c79501d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L42
+
+### SAN SNI validation
+The server certificate validation described in [A29 SAN matching][A29_SAN-matching]
+matches the Subject Alternative Names specified in the server certificate against 
+[`match_subject_alt_names`][match_subject_alt_names] in `CertificateValidationContext`.
+If `auto_sni_san_validation` is set in the [UpstreamTlsContext][UTC], matching will be 
+performed against the SNI that was used by the client, and this validation will replace
+the [`match_subject_alt_names`][match_subject_alt_names] if set. This verification occurs
+in the TrustManager of the SslContext which is created using the cert store indicated by 
+`CertificateValidationContext` in `UpstreamTlsContext` which is either a managed cert store
+or the system root cert store. 
+
+#### Caching for the SslContext 
+The `SslContextProviderSupplier` (named so because it supplies both client and server
+SslContext providers) creates a provider for the client `SslContext` and today 
+maintains a cache of `UpstreamTlsContext` to the client `SslContext` provider instances. 
+For the SNI requirement, the `TrustManager` in the `SslContext` needs to 
+be aware of the SNI to validate the SAN against, so a different `TrustManager` instance needs 
+to be created for each SNI to use for the same `UpstreamTlsContext`, so this cache's key will 
+need to be enhanced to be <UpstreamTlsContext, String> to hold the SNI as well, and the client
+`SslContext` provider for a particular key will create a `TrustManager` instance that takes the 
+SNI to validate the SANs against and set it in the `SslContext` it provides.
+
+[A29_SAN-matching]: https://github.com/grpc/proposal/blob/master/A29-xds-tls-security.md#server-authorization-aka-subject-alt-name-checks
```

Comment:
Don't need the "https://github.com/grpc/proposal/blob/master/" part.  This should be a relative link.